dtm0/recovery.c

```diff
@@ -1677,11 +1731,40 @@ static void dtm0_restore(struct m0_fom        *fom,
 
 	do {
 		M0_SET0(&record);
-		rc = recovery_machine_log_iter_next(rf->rf_m, &rf->rf_log_iter,
-						    &rf->rf_tgt_svc, NULL,
-						    &record);
-		/* Any value except zero means that we should stop recovery. */
-		F(next) = rc == 0;
+		if (F(last_dtx_met)) {
+			/*
+			 * Last iteration reached the RECOVERING mark in the
+			 * log, don't need to send REDOs past this mark.
+			 */
+			F(next) = false;
+		} else {
+			do {
+				rc = recovery_machine_log_iter_next(
+						rf->rf_m, &rf->rf_log_iter,
+						&record);
+				/*
+				 * Any value except zero means that we should
+				 * stop recovery.
+				 *
+				 * XXX: handle cases when rc not in (0,
+				 * -ENOENT).
+				 */
+				F(next) = (rc == 0);
+				if (!F(next))
+					break;
+				rc = recovery_machine_dtxid_cmp(
+						rf->rf_m,
+						&record.dlr_txd.dtd_id,
+						&F(last_dtx_id));
+				F(last_dtx_met) = (rc == 0);
+				if (participated(&record, &rf->rf_tgt_svc))
+					break;
+				m0_dtm0_log_iter_rec_fini(&record);
+				if (F(last_dtx_met))
+					F(next) = false;
+			} while (!F(last_dtx_met));
```

Comment:
yeah good catch.

`m0_dtm0_log_iter_rec_fini` internaly resets the record to all-zeroes (indirectly).  So there is no bug in the code now.  Should I set it to all-zeroes explicitly, e.g. `M0_SET0(&record);` ?

Reply:
Yes, in addition to fini() (not instead of it).

Reply:
Agree with @andriytk . At least we should add comments to explain there is no issue to access the `record` there.

Reply:
The suggestion from Andriy above is not exactly correct, last_dtx_met does not mean we need to bzero the record.  I'll put M0_SET0 at the right place inside the loop:

```
diff --git a/dtm0/recovery.c b/dtm0/recovery.c
index 824471ea..020091a2 100644
--- a/dtm0/recovery.c
+++ b/dtm0/recovery.c
@@ -1760,11 +1762,21 @@ static void dtm0_restore(struct m0_fom        *fom,
                                if (participated(&record, &rf->rf_tgt_svc))
                                        break;
                                m0_dtm0_log_iter_rec_fini(&record);
+                               M0_SET0(&record);
                                if (F(last_dtx_met))
                                        F(next) = false;
                        } while (!F(last_dtx_met));
                }
```

Reply:
OK, but I still think it would be better to do it after the loop (more explicit and clear):

```c
if (!F(next))
    M0_SET0(&record);
```

Reply:
Ok, since SET0 performance is a concern -- changing the code to get rid of SET0/IF0 altogether.  Note that DTM0 record includes `dd_ancient_dtx` which is huge (1k), so concern about performance is valid.

```diff
diff --git a/dtm0/recovery.c b/dtm0/recovery.c
index b2ad5e2d..9bf94d3d 100644
--- a/dtm0/recovery.c
+++ b/dtm0/recovery.c
@@ -1740,7 +1740,6 @@ static void dtm0_restore(struct m0_fom        *fom,
         * need both flags to define what to do.
         */
        do {
-               M0_SET0(&record);
                if (F(last_dtx_met)) {
                        /*
                         * Last iteration reached the RECOVERING mark in the
@@ -1762,7 +1761,6 @@ static void dtm0_restore(struct m0_fom        *fom,
                                 */
                                M0_ASSERT(M0_IN(rc, (0, -ENOENT)));
                                F(next) = (rc == 0);
-                               M0_ASSERT(ergo(rc == -ENOENT, M0_IS0(&record)));
                                if (!F(next))
                                        break;
                                rc = recovery_machine_dtxid_cmp(
@@ -1773,7 +1771,6 @@ static void dtm0_restore(struct m0_fom        *fom,
                                if (participated(&record, &rf->rf_tgt_svc))
                                        break;
                                m0_dtm0_log_iter_rec_fini(&record);
-                               M0_SET0(&record);
                                if (F(last_dtx_met))
                                        F(next) = false;
                        } while (!F(last_dtx_met));
@@ -1781,20 +1778,21 @@ static void dtm0_restore(struct m0_fom        *fom,

                /*
                 * At this point, a record is either fully initialized and must
-                * be sent, or is all-zero and F(next) is then set to false, so
-                * we will send EOL flag.  With current implementation, EOL
-                * flags should be sent on a separate message, which must not
-                * have any payload.
+                * be sent, or is holding garbage from previous iterations and
+                * F(next) is then set to false, so we will send EOL flag.  With
+                * current implementation, EOL flags should be sent on a
+                * separate message, which must not have any payload.
                 */
-               M0_ASSERT(equi(M0_IS0(&record), !F(next)));
-
                redo = (struct dtm0_req_fop) {
                        .dtr_msg       = DTM_REDO,
                        .dtr_initiator = F(initiator),
-                       .dtr_payload   = record.dlr_payload,
-                       .dtr_txr       = record.dlr_txd,
-                       .dtr_flags     = F(next) ? 0 : M0_BITS(M0_DMF_EOL),
                };
+               if (F(next)) {
+                       redo.dtr_payload   = record.dlr_payload;
+                       redo.dtr_txr       = record.dlr_txd;
+               } else {
+                       redo.dtr_flags     = M0_BITS(M0_DMF_EOL);
+               }

                /*
                 * TODO: there is extra memcpy happening here -- first copy done
```